north/src/runtime/npk/linux.rs

```diff
@@ -137,62 +138,47 @@ async fn install_internal(
     dm: &dm::Dm,
     lc: &LoopControl,
     npk: &Path,
-) -> Result<(Name, Version)> {
+) -> std::result::Result<(Name, Version), InstallFailure> {
     let start = time::Instant::now();
-
     if let Some(npk_name) = npk.file_name() {
-        info!("Loading {}", npk_name.to_string_lossy());
+        info!(
+            "Installing {}, loading npk file",
+            npk_name.to_string_lossy()
+        );
     }
 
-    let file =
-        std::fs::File::open(&npk).with_context(|| format!("Failed to open {}", npk.display()))?;
-    let reader = std::io::BufReader::new(file);
-    let mut archive = zip::ZipArchive::new(reader).context("Failed to read zip")?;
-
-    debug!("Loading hashes");
-    let hashes = {
-        let mut signature_file = archive
-            .by_name(SIGNATURE)
-            .with_context(|| format!("Failed to read signature from {}", npk.display()))?;
-        let mut signature = String::new();
-        signature_file.read_to_string(&mut signature)?;
-        Hashes::from_str(&signature, &state.signing_keys).context("Failed to load hashes")?
-    };
-
-    let manifest = {
-        let mut manifest_file = archive
-            .by_name(MANIFEST)
-            .with_context(|| format!("Failed to read manifest from {}", npk.display()))?;
-        let mut manifest = String::new();
-        manifest_file.read_to_string(&mut manifest)?;
-        let digest = sha2::Sha256::digest(manifest.as_bytes());
-        if hex::decode(&hashes.manifest_hash)? != digest.as_slice() {
-            return Err(anyhow!("Invalid manifest hash"));
-        }
-        Manifest::from_str(&manifest)?
-    };
+    debug!(
+        "npk size: {}",
+        std::fs::metadata(&npk)
+            .expect("Could not get metadata of file")
+            .len()
+    );
+    let p: &std::path::Path = npk.into();
+    let mut archive_reader = ArchiveReader::new(&p, &state.signing_keys)?;
+
+    let hashes = archive_reader.extract_hashes()?;
+
+    let manifest = archive_reader.extract_manifest_from_archive()?;
     debug!("Manifest loaded for \"{}\"", manifest.name);
+
```

Comment:
Here we should check if we have a container with `name` and `version` already installed.

Reply:
is checked on state level...should not be part of platform specific code